### ISSUE TITLE
Upload HH Interview Spreadsheet

### ISSUE THREAD

**Issue by daiClegg** [dbrownukk/EFD_v2#210]
Upload HH Interview (also Parse & Validate) should work similarly to OHEA Upload of WG Interview.
The Spreadsheet format is different but has much in common.

One additional validation rule is that the GPS property of the Study should be non-null or Upload should not be permitted. This is to enforce the best practice of collecting GPS

**Comment by daiClegg** [comment]
See also #128 for treatment of new assets in the Resources tabs of HH Interview

**Comment by daiClegg** [comment]
This feature will behave exactly like the equivalent feature for uploading WG Interview data in OHEA.
This applies to the UX and to the state lifecycle of the HH data set.
The content is different according to the differences in the two uploaded spreadsheets and the equivalent data models.

The issue is current;y marked as blocked by the HH Generate (#211)  but in fact, it is only blocked by having the definition of the spreadsheet agreed.

**Comment by daiClegg** [comment]
Design data model added in:

 https://github.com/dbrownukk/EFD_v2/blob/master/DesignDoc/IHM%26HEADesignModel7.mwb

**Comment by daiClegg** [comment]
This is ready for testing no?
I assume so and will add to testing, since #222 is testable

**Comment by daiClegg** [comment]
Testing will proceed with dummy data, but we need real-world example data to have any confidence.
@celiapetty i think i already requested some real data. Can you upload here please?
If you sent it to me in mail apologies i can't find it.

**Comment by celiapetty** [comment]
Do you need a spreadsheet? Apologies initial request must have come when I was away

**Comment by daiClegg** [comment]
Initial test note
--
Instance: EFD_HM
Study:  M1 Workflow Study 2018
HH: HH1

- i like the error message for non-matching study in the database & on the spreadsheet - good quality check, but it doesn't stop parsing so should it be a warning message (amber)?

- The spreadsheet has 3 HH members but only one is loaded.

- HH Member  standard questions:  
 -- Absence is a Yes/No; the spreadsheet has it as a number of months.
 -- The spreadsheet does not include the 'Reason for Absence' question, but the form has space for an answer

- HH Member configured questions
 -- Spreadsheet has one which is absent from the form

- Asset cash (UGX) was set valid in parse, but the Name was left black - which threw an error on save (correctly)
I assigned UGX to check it was there, and now i can't unassign it to restore the parse output

- Livestock product 2 (Goat Meat) did not parse
Goat meat is a valid livestock product.
I  assigned it to check and i can't unassign it to restore the parse output
(Bert's Place has the same data - uncorrected)

- Why is Employment 2 (Maid) invalid? 
There is part payment in cash and part payment in food.
I think that should be allowable

- Inputs Fertilizer  parses as correct, but the Name field is blank

 - Inputs  Name dropdown shows Wild Foods; it should show Non-Food Purchases and Other Tradable Goods

**Comment by daiClegg** [comment]
@celiapetty the spreadsheet i use for testing is this one.

[UploadTest160319.xlsx](https://github.com/dbrownukk/EFD_v2/files/2980706/UploadTest160319.xlsx)

Please can you review it and add some more realistic data.
Feel free to include invalid data, but put a note in the spreadsheet in one of the spare blank rows to indicate - so i'll know to expect a failure.

Don't bother with the first 3 tabs, just concentrate on the Assets and other resources.
Thanks

**Comment by daiClegg** [comment]
Overall this is looking very good for a first test; the workflow, echoing OHEA closely, works fine.
Nice.

**Comment by celiapetty** [comment]
Spreadsheet with real data. I have left dummy data where data in 'real' interview doesn't cover the point eg labour paid in food
[UploadTest160319+CP.xlsx](https://github.com/dbrownukk/EFD_v2/files/2985548/UploadTest160319%2BCP.xlsx)

**Comment by daiClegg** [comment]
This is great @celiapetty . Thanks.

I was able to load and parse this with no problems.
There are lots of items, though that don't match Resource SubTypes (RST) in the database.
Some, i see, you have introduced as new synonyms, some as (deliberate?) typos. 
And some are genuine omissions
I will create synonyms & new resource subtypes as i see fit since this is just for testing.
We should review all the RSTs before production but that's a few months away i guess.

**Comment by daiClegg** [comment]
Uploaded and validate the Real World test data spreadsheet.
I had to create plenty of new RSTs to support this -  but that's fine.

- One outstanding issue (as noted above) is that the LOV for Inputs should be a concatenation of Non-Food Purchases & Other Tradable Goods; it is currently Wild Foods

- The other issue (as discussed) is that the fix for Issue #242 means i can't save a partially corrected corrected spreadsheet parse.

Looks very good, modulo the anomalies noted above.

**Comment by dbrownukk** [comment]
Copy of above issues so that I can track fixing...

Instance: EFD_HM
Study: M1 Workflow Study 2018
HH: HH1

1. i like the error message for non-matching study in the database & on the spreadsheet - good quality check, but it doesn't stop parsing so should it be a warning message (amber)?

** Done **

2. The spreadsheet has 3 HH members but only one is loaded.
HH Member  standard questions:

** Done ** - Tested in M1 Workflow Study and all 3 HHM are uploaded 

3. -- Absence is a Yes/No; the spreadsheet has it as a number of months.

-- The spreadsheet does not include the 'Reason for Absence' question, but the form has space for an answer
HH Member configured questions

Now have:

Absent Y/N
Reason for Absence - text
Period away in Months - Integer





-- Spreadsheet has one which is absent from the form
Asset cash (UGX) was set valid in parse, but the Name was left black - which threw an error on save (correctly)

I assigned UGX to check it was there, and now i can't unassign it to restore the parse output
Livestock product 2 (Goat Meat) did not parse

Goat meat is a valid livestock product.

I  assigned it to check and i can't unassign it to restore the parse output

(Bert's Place has the same data - uncorrected)
Why is Employment 2 (Maid) invalid?

There is part payment in cash and part payment in food.

I think that should be allowable
Inputs Fertilizer  parses as correct, but the Name field is blank
Inputs  Name dropdown shows Wild Foods; it should show Non-Food Purchases and Other Tradable Goods

**Comment by daiClegg** [comment]
Why is Employment 2 (Maid) invalid?

> 'Unit' needs populating in spreadsheet. 
There is part payment in cash and part payment in food.

> I think that should be allowable
> 
> ** Should or should not? ** Currently is allowable

That's fine, it whoulc be allowed. 
I hadn't noticed the invalid data.
It would be nice to have a more user-friendly error message.

**Comment by dbrownukk** [comment]
Think we are there with Upload and Parse. 

As we get some more real world data lets raise new issues as bugs. 

Does that work?

**Comment by daiClegg** [comment]
I agree it's feature complete.
Will re-test and close; if any outstanding anomalies will raise as separate issues.

**Comment by daiClegg** [comment]
i like the error message for non-matching study in the database & on the spreadsheet - good quality check, but it doesn't stop parsing so should it be a warning message (amber)?
** Done **

**Tested OK**

**Comment by daiClegg** [comment]
The spreadsheet has 3 HH members but only one is loaded.
HH Member standard questions:

** Done ** - Tested in M1 Workflow Study and all 3 HHM are uploaded

placeholder comment for next test; blocked by other issues right now...

**Comment by daiClegg** [comment]
-- Absence is a Yes/No; the spreadsheet has it as a number of months.
-- The spreadsheet does not include the 'Reason for Absence' question, but the form has space for an answer
HH Member configured questions

Now have:

Absent Y/N
Reason for Absence - text
Period away in Months - Integer

Parses correctly into HHM

**Tested Ok.**

**Comment by daiClegg** [comment]
4. Asset cash (UGX) was set valid in parse, but the Name was left black - which threw an error on save (correctly)

I assigned UGX to check it was there, and now i can't unassign it to restore the parse output

** Fixed **

Cash Assets now match currencies correctly.

**Tested Ok**

**Comment by daiClegg** [comment]
Livestock product 2 (Goat Meat) did not parse
Goat meat is a valid livestock product.

I assigned it to check and i can't unassign it to restore the parse output

LSP resourcetype has a resource subtype of Goats meat. It also has a synonym of 'Meat' which is a 'Food Purchase' . Need a rule for this scenario..

** Fixed ** by ignoring synonym check in checksubtype()

(Bert's Place has the same data - uncorrected)

Study: M1 End2End Test
HH:      3

**Tested Ok**

**Comment by daiClegg** [comment]
All previous test anomalies now retested ok or legged separatly 
This Feature can be regarded as Feature Complete and closed
